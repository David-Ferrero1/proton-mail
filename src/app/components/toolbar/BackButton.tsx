import React from 'react';
import { Icon } from 'react-components';

import ToolbarButton from './ToolbarButton';

interface Props {
    onClick: () => void;
}

const BackButton = ({ onClick }: Props) => {
    return (
        <ToolbarButton onClick={onClick}>
            <Icon className="toolbar-icon mauto" name="arrow-left" />
        </ToolbarButton>
    );
};

export default BackButton;
