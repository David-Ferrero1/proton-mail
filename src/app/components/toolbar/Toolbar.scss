// TODO: Remove this in favor of something clean in the DS

@import '~design-system/_sass/reusable-components/_design-system-config.scss';

.toolbar-button {
    color: white;
}

.toolbar-dropdown {
    ul {
        margin: 0;
    }

    button:disabled {
        opacity: .5
    }
}
