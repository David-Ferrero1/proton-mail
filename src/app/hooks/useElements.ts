import { useState, useEffect, useMemo } from 'react';
import { useApi, useCache } from 'react-components';
import { queryConversations, getConversation } from 'proton-shared/lib/api/conversations';
import { queryMessageMetadata, getMessage } from 'proton-shared/lib/api/messages';
import { EVENT_ACTIONS } from 'proton-shared/lib/constants';
import { toMap } from 'proton-shared/lib/helpers/object';

import { Conversation } from '../models/conversation';
import { sort as sortElements, hasLabel, parseLabelIDsInEvent, isSearch, isFilter } from '../helpers/elements';
import { Element } from '../models/element';
import { Page, Filter, Sort, SearchParameters } from '../models/tools';
import { expectedPageLength } from '../helpers/paging';
import {
    ElementEvent,
    Event,
    ElementCountEvent,
    ConversationEvent,
    MessageEvent,
    LabelIDsChanges
} from '../models/event';
import { useSubscribeEventManager } from './useHandler';
import { useExpirationCheck } from './useExpiration';
import { ConversationCountsModel, MessageCountsModel } from 'proton-shared/lib/models';

interface Options {
    conversationMode: boolean;
    labelID: string;
    page: Page;
    sort: Sort;
    filter: Filter;
    search: SearchParameters;
}

interface CacheParams {
    labelID: string;
    sort: Sort;
    filter: Filter;
    address?: string;
    from?: string;
    to?: string;
    keyword?: string;
    begin?: number;
    end?: number;
    attachments?: number;
    wildcard?: number;
}

interface Cache {
    params: CacheParams;
    page: Page;
    pages: number[];
    elements: { [ID: string]: Element };
    updatedElements: string[];
}

const emptyCache = (page: Page, params: CacheParams): Cache => ({
    params,
    page,
    elements: {},
    pages: [],
    updatedElements: []
});

export const useElements = ({
    conversationMode,
    labelID,
    search,
    page,
    sort,
    filter
}: Options): [string, Conversation[], boolean, number] => {
    const api = useApi();
    const [loading, setLoading] = useState(false);
    const [invalidated, setInvalidated] = useState(false);
    const [beforeFirstLoad, setBeforeFirstLoad] = useState(true);
    const [localCache, setLocalCache] = useState<Cache>(
        emptyCache(page, {
            labelID,
            sort,
            filter,
            ...search
        })
    );
    // Warning: this hook relies mainly on a localCache, not the globalCache
    // This import is needed only for a specific use case (message expiration)
    const globalCache = useCache();

    // Remove from cache expired elements
    useExpirationCheck(Object.values(localCache.elements), (element) => {
        const elements = Object.keys(localCache.elements).reduce<{ [ID: string]: Element }>((acc, cacheID) => {
            if (element.ID !== cacheID) {
                acc[cacheID] = localCache.elements[cacheID];
            }
            return acc;
        }, {});

        setLocalCache({ ...localCache, elements });

        globalCache.delete(ConversationCountsModel.key);
        globalCache.delete(MessageCountsModel.key);
    });

    // Compute the conversations list from the cache
    const elements = useMemo(() => {
        // Getting all params from the cache and not from scoped params
        // To prevent any desynchronization between cache and the output of the memo
        const {
            params: { labelID, sort },
            page
        } = localCache;

        const minPage = localCache.pages.reduce((acc, page) => (page < acc ? page : acc), localCache.pages[0]);
        const startIndex = (page.page - minPage) * page.size;
        const endIndex = startIndex + page.size;
        const elementsArray = Object.values(localCache.elements);
        const filtered = elementsArray.filter((element) => hasLabel(element, labelID));
        const sorted = sortElements(filtered, sort, labelID);
        return sorted.slice(startIndex, endIndex);
    }, [localCache]);

    const total = useMemo(() => localCache.page.total, [localCache.page.total]);

    const paramsChanged = () =>
        labelID !== localCache.params.labelID ||
        sort !== localCache.params.sort ||
        filter !== localCache.params.filter ||
        search.address !== localCache.params.address ||
        search.from !== localCache.params.from ||
        search.to !== localCache.params.to ||
        search.keyword !== localCache.params.keyword ||
        search.begin !== localCache.params.begin ||
        search.end !== localCache.params.end ||
        search.attachments !== localCache.params.attachments ||
        search.wildcard !== localCache.params.wildcard;

    const pageCached = () => localCache.pages.includes(page.page);

    const pageChanged = () => localCache.page.page !== page.page;

    const pageIsConsecutive = () =>
        localCache.pages.some((p) => p === page.page || p === page.page - 1 || p === page.page + 1);

    const hasListFromTheStart = () => localCache.pages.includes(0);

    const lastHasBeenUpdated = () =>
        elements.length === page.size &&
        page.page === Math.max.apply(null, localCache.pages) &&
        localCache.updatedElements.includes(elements[elements.length - 1].ID || '');

    // Live cache means we listen to events from event manager without refreshing the list every time
    const isLiveCache = () => !isSearch(search) && !isFilter(filter) && hasListFromTheStart();

    const isExpectedLength = () => !isLiveCache() || elements.length === expectedPageLength({ ...page, total });

    const shouldResetCache = () =>
        !loading && (invalidated || paramsChanged() || !pageIsConsecutive() || lastHasBeenUpdated());

    const shouldSendRequest = () => !loading && (shouldResetCache() || !pageCached() || !isExpectedLength());

    const shouldUpdatePage = () => pageChanged() && pageCached();

    const updatePage = () => {
        setLocalCache({
            ...localCache,
            page: { ...localCache.page, page: page.page }
        });
    };

    const queryElement = async (elementID: string): Promise<Element> => {
        const query = conversationMode ? getConversation : getMessage;
        const result: any = await api(query(elementID));
        return conversationMode ? result.Conversation : result.Message;
    };

    const queryElements = async (): Promise<{ Total: number; Elements: Element[] }> => {
        const query = conversationMode ? queryConversations : queryMessageMetadata;
        const result: any = await api(
            query({
                Page: page.page,
                PageSize: page.size,
                Limit: page.limit,
                LabelID: labelID,
                Sort: sort.sort,
                Desc: sort.desc ? 1 : 0,
                Begin: search.begin,
                End: search.end,
                // BeginID,
                // EndID,
                Keyword: search.keyword,
                To: search.to,
                From: search.from,
                // Subject,
                Attachments: search.attachments,
                Unread: filter.Unread,
                AddressID: search.address,
                // ID,
                AutoWildcard: search.wildcard
            } as any)
        );

        return {
            Total: result.Total,
            Elements: conversationMode ? result.Conversations : result.Messages
        };
    };

    const resetCache = () =>
        setLocalCache(
            emptyCache(page, {
                labelID,
                sort,
                filter,
                ...search
            })
        );

    const load = async () => {
        setLoading(true);
        try {
            const { Total, Elements } = await queryElements();
            const elementsMap = toMap(Elements, 'ID');
            const updatedElements = localCache.updatedElements.filter((elementID) => !elementsMap[elementID]);
            setInvalidated(false);
            setLocalCache(
                (localCache: Cache): Cache => {
                    return {
                        params: localCache.params,
                        page: {
                            ...localCache.page,
                            page: page.page,
                            total: Total
                        },
                        pages: [...localCache.pages, page.page],
                        elements: {
                            ...localCache.elements,
                            ...elementsMap
                        },
                        updatedElements
                    };
                }
            );
        } finally {
            setLoading(false);
        }
    };

    // Main effect watching all inputs and responsible to trigger actions on the cache
    useEffect(() => {
        shouldResetCache() && resetCache();
        shouldSendRequest() && load();
        shouldUpdatePage() && updatePage();
        setBeforeFirstLoad(false);
    }, [
        labelID,
        page,
        sort,
        filter,
        search.address,
        search.from,
        search.to,
        search.keyword,
        search.begin,
        search.end,
        search.attachments,
        search.wildcard,
        invalidated,
        localCache.updatedElements
    ]);

    // Listen to event manager and update de cache
    useSubscribeEventManager(
        async ({ Conversations = [], Messages = [], ConversationCounts = [], MessageCounts = [] }: Event) => {
            const Elements: ElementEvent[] = conversationMode ? Conversations : Messages;
            const Counts: ElementCountEvent[] = conversationMode ? ConversationCounts : MessageCounts;

            if (!isLiveCache() && Elements.length) {
                setInvalidated(true);
                return;
            }

            const count = Counts.find((count) => count.LabelID === labelID);

            const { toDelete, toUpdate, toCreate } = Elements.reduce(
                (acc, event) => {
                    const { ID, Action } = event;
                    const Element = conversationMode
                        ? (event as ConversationEvent).Conversation
                        : (event as MessageEvent).Message;

                    switch (Action) {
                        case EVENT_ACTIONS.DELETE:
                            acc.toDelete.push(ID);
                            break;
                        case EVENT_ACTIONS.UPDATE_DRAFT:
                        case EVENT_ACTIONS.UPDATE_FLAGS:
                            acc.toUpdate.push({ ID, ...Element });
                            break;
                        case EVENT_ACTIONS.CREATE:
                            acc.toCreate.push(Element);
                            break;
                    }
                    return acc;
                },
                {
                    toDelete: [] as string[],
                    toUpdate: [] as (Element & LabelIDsChanges)[],
                    toCreate: [] as (Element & LabelIDsChanges)[]
                }
            );

            const toUpdateCompleted = await Promise.all(
                toUpdate.map(async (element) => {
                    const elementID = element.ID || '';
                    const existingElement = localCache.elements[elementID];

                    if (existingElement) {
                        element = parseLabelIDsInEvent(existingElement, element);
                    }

                    return existingElement ? { ...existingElement, ...element } : queryElement(elementID);
                })
            );

            setLocalCache((localCache) => {
                const newReplacements: { [ID: string]: Element } = {};

                [...toCreate, ...toUpdateCompleted].forEach((element) => {
                    newReplacements[element.ID || ''] = element;
                });
                const newElements = {
                    ...localCache.elements,
                    ...newReplacements
                };
                toDelete.forEach((elementID) => {
                    delete newElements[elementID];
                });

                const updatedElements = [...localCache.updatedElements, ...Object.keys(newReplacements), ...toDelete];

                return {
                    ...localCache,
                    elements: newElements,
                    page: {
                        ...localCache.page,
                        total: count ? count.Total : localCache.page.total
                    },
                    updatedElements
                };
            });
        }
    );

    return [localCache.params.labelID, elements, beforeFirstLoad || loading, localCache.page.total];
};
