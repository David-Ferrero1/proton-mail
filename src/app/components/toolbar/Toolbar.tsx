import React, { useMemo } from 'react';
import { c } from 'ttag';
import { Icon, useLabels, useFolders } from 'react-components';
import { identity } from 'proton-shared/lib/helpers/function';
import { MailSettings } from 'proton-shared/lib/interfaces';

import ToolbarSeparator from './ToolbarSeparator';
import ReadUnreadButtons from './ReadUnreadButtons';
import ToolbarDropdown from './ToolbarDropdown';
import LayoutDropdown from './LayoutDropdown';
import MoveButtons from './MoveButtons';
import DeleteButton from './DeleteButton';
import EmptyButton from './EmptyButton';
import SortDropdown from './SortDropdown';
import FilterDropdown from './FilterDropdown';
import SelectAll from './SelectAll';
import MoveDropdown from '../dropdown/MoveDropdown';
import LabelDropdown from '../dropdown/LabelDropdown';
import BackButton from './BackButton';
import PagingControls from './PagingControls';
import { Page, Sort, Filter } from '../../models/tools';
import { Element } from '../../models/element';
import { Breakpoints } from '../../models/utils';
import NavigationControls from './NavigationControls';

interface Props {
    loading?: boolean;
    onCheck: (IDs: string[], checked: boolean, replace: boolean) => void;
    labelID: string;
    elementID?: string;
    elements: Element[];
    selectedIDs: string[];
    mailSettings: MailSettings;
    columnMode: boolean;
    conversationMode: boolean;
    breakpoints: Breakpoints;
    page: Page;
    onPage: (page: number) => void;
    sort: Sort;
    onSort: (sort: Sort) => void;
    filter: Filter;
    onFilter: (filter: Filter) => void;
    onBack: () => void;
    onElement: (element: Element) => void;
    onNavigate: (labelID: string) => void;
}

const Toolbar = ({
    labelID = '',
    elementID,
    elements,
    onCheck,
    mailSettings,
    columnMode,
    conversationMode,
    breakpoints,
    selectedIDs = [],
    loading = false,
    onSort,
    sort,
    onFilter,
    filter,
    onBack,
    page,
    onPage,
    onElement,
    onNavigate
}: Props) => {
    const [labels] = useLabels();
    const [folders] = useFolders();
    const selectedElements = useMemo(
        () =>
            selectedIDs
                .map((elementID) => elements.find((element) => element.ID === elementID) as Element)
                .filter(identity),
        [elements, selectedIDs]
    );
    const listInView = columnMode || !elementID;

    return (
        <nav className="toolbar toolbar--heavy flex noprint flex-spacebetween">
            <div className="flex">
                {listInView ? (
                    <SelectAll
                        labelID={labelID}
                        elements={elements}
                        selectedIDs={selectedIDs}
                        onCheck={onCheck}
                        loading={loading}
                    />
                ) : (
                    <BackButton onClick={onBack} />
                )}
                <ToolbarSeparator />
                <ReadUnreadButtons
                    mailSettings={mailSettings}
                    selectedIDs={selectedIDs}
                    onBack={onBack}
                    labelID={labelID}
                    elements={elements}
                />
                <ToolbarSeparator />
                <MoveButtons
                    labelID={labelID}
                    labels={labels}
                    folders={folders}
                    breakpoints={breakpoints}
                    selectedIDs={selectedIDs}
                    onBack={onBack}
                />
                <DeleteButton
                    labelID={labelID}
                    conversationMode={conversationMode}
                    breakpoints={breakpoints}
                    selectedIDs={selectedIDs}
                />
                <ToolbarSeparator />
                <ToolbarDropdown
                    autoClose={false}
                    noMaxSize={true}
                    disabled={!selectedElements || !selectedElements.length}
                    content={<Icon className="toolbar-icon" name="folder" />}
                    dropDownClassName="moveDropdown"
                    title={c('Title').t`Move to`}
                    data-test-id="toolbar:moveto"
                >
                    {({ onClose, onLock }) => (
                        <MoveDropdown
                            labelID={labelID}
                            elements={selectedElements}
                            conversationMode={conversationMode}
                            onClose={onClose}
                            onLock={onLock}
                            onBack={onBack}
                            breakpoints={breakpoints}
                        />
                    )}
                </ToolbarDropdown>
                <ToolbarDropdown
                    autoClose={false}
                    noMaxSize={true}
                    disabled={!selectedElements || !selectedElements.length}
                    content={<Icon className="toolbar-icon" name="label" />}
                    dropDownClassName="labelDropdown"
                    title={c('Title').t`Label as`}
                    data-test-id="toolbar:labelas"
                >
                    {({ onClose, onLock }) => (
                        <LabelDropdown
                            labelID={labelID}
                            labels={labels}
                            elements={selectedElements}
                            onClose={onClose}
                            onLock={onLock}
                            breakpoints={breakpoints}
                        />
                    )}
                </ToolbarDropdown>
                <EmptyButton labelID={labelID} breakpoints={breakpoints} elements={elements} />
            </div>
            <div className="flex">
                {breakpoints.isDesktop && (
                    <>
                        {listInView && (
                            <>
                                <FilterDropdown
                                    labelID={labelID}
                                    loading={loading}
                                    filter={filter}
                                    onFilter={onFilter}
                                    onNavigate={onNavigate}
                                />
                                <SortDropdown
                                    loading={loading}
                                    conversationMode={conversationMode}
                                    sort={sort}
                                    onSort={onSort}
                                />
                            </>
                        )}
                        <LayoutDropdown mailSettings={mailSettings} />
                        <ToolbarSeparator />
                    </>
                )}
                {listInView ? (
                    <PagingControls loading={loading} page={page} onPage={onPage} />
                ) : (
                    <NavigationControls
                        loading={loading}
                        conversationMode={conversationMode}
                        elementID={elementID}
                        elements={elements}
                        onElement={onElement}
                    />
                )}
            </div>
        </nav>
    );
};

export default Toolbar;
