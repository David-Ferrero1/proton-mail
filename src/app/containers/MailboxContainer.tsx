import React, { useState, useMemo, useEffect } from 'react';
import { PrivateMainArea } from 'react-components';
import { History, Location } from 'history';
import { MailSettings, UserSettings } from 'proton-shared/lib/interfaces';
import { getSearchParams } from 'proton-shared/lib/helpers/url';

import { Element } from '../models/element';
import { Sort, Filter, Page, SearchParameters } from '../models/tools';
import { useMailboxPageTitle } from '../hooks/useMailboxPageTitle';
import { useElements } from '../hooks/useElements';
import { isColumnMode, isConversationMode } from '../helpers/mailSettings';
import {
    pageFromUrl,
    sortFromUrl,
    filterFromUrl,
    setPageInUrl,
    setSortInUrl,
    setFilterInUrl,
    setPathInUrl,
    extractSearchParameters
} from '../helpers/mailboxUrl';
import Toolbar from '../components/toolbar/Toolbar';
import List from '../components/list/List';
import ConversationView from '../components/conversation/ConversationView';
import PlaceholderView from '../components/view/PlaceholderView';
import MessageOnlyView from '../components/message/MessageOnlyView';
import { PAGE_SIZE } from '../constants';
import { isMessage, isSearch as testIsSearch } from '../helpers/elements';
import { isDraft } from '../helpers/message/messages';
import { Message } from '../models/message';
import { Breakpoints } from '../models/utils';
import { OnCompose } from '../hooks/useCompose';
import { useWelcomeFlag } from '../hooks/useWelcomeFlag';
import useNewEmailNotification from '../hooks/useNewEmailNotification';

import './MailboxContainer.scss';

interface Props {
    labelID: string;
    userSettings: UserSettings;
    mailSettings: MailSettings;
    breakpoints: Breakpoints;
    elementID?: string;
    location: Location;
    history: History;
    onCompose: OnCompose;
    throughLogin: boolean;
}

const MailboxContainer = ({
    labelID: inputLabelID,
    userSettings,
    mailSettings,
    breakpoints,
    elementID,
    location,
    history,
    onCompose,
    throughLogin
}: Props) => {
    const forceRowMode = breakpoints.isNarrow || breakpoints.isTablet;
    const columnModeSetting = isColumnMode(mailSettings);
    const columnMode = columnModeSetting && !forceRowMode;
    const columnLayout = columnModeSetting || forceRowMode;

    // Page state is hybrid: page number is handled by the url, total computed in useElements, size and limit are constants
    // Yet, it is simpler to co-localize all these data in one object
    const [page, setPage] = useState<Page>({
        page: pageFromUrl(location),
        total: 0,
        size: PAGE_SIZE,
        limit: PAGE_SIZE
    });

    const searchParams = getSearchParams(location.search);
    const conversationMode = isConversationMode(inputLabelID, mailSettings, location);
    const searchParameters = useMemo<SearchParameters>(() => extractSearchParameters(location), [
        searchParams.address,
        searchParams.from,
        searchParams.to,
        searchParams.keyword,
        searchParams.begin,
        searchParams.end,
        searchParams.attachments,
        searchParams.wildcard
    ]);
    const isSearch = testIsSearch(searchParameters);
    const sort = useMemo<Sort>(() => sortFromUrl(location), [searchParams.sort]);
    const filter = useMemo<Filter>(() => filterFromUrl(location), [searchParams.filter]);

    const [checkedElements, setCheckedElements] = useState<{ [ID: string]: boolean }>({});
    useNewEmailNotification(history, location);
    const { labelID, elements, loading, expectedLength, total } = useElements({
        conversationMode,
        labelID: inputLabelID,
        page,
        sort,
        filter,
        search: searchParameters
    });

    useEffect(() => setPage({ ...page, page: pageFromUrl(location) }), [searchParams.page]);
    useEffect(() => setPage({ ...page, total }), [total]);
    useEffect(() => setCheckedElements({}), [labelID]);

    useMailboxPageTitle(labelID, location);

    const checkedIDs = useMemo(() => {
        return Object.entries(checkedElements).reduce((acc, [elementID, isChecked]) => {
            if (!isChecked) {
                return acc;
            }
            acc.push(elementID);
            return acc;
        }, [] as string[]);
    }, [checkedElements]);

    const selectedIDs = useMemo(() => {
        if (checkedIDs.length) {
            return checkedIDs;
        }
        if (elementID) {
            return [elementID];
        }
        return [];
    }, [checkedIDs, elementID]);

    const welcomeFlag = useWelcomeFlag(throughLogin, [labelID, selectedIDs.length]);

    const handleElement = (element: Element) => {
        history.push(setPathInUrl(location, labelID, element.ID));
        if (isMessage(element) && isDraft(element)) {
            onCompose({ existingDraft: { localID: element.ID as string, data: element as Message } });
        }
        setCheckedElements({});
    };
    const handleBack = () => history.push(setPathInUrl(location, labelID));
    const handlePage = (pageNumber: number) => history.push(setPageInUrl(location, pageNumber));
    const handleSort = (sort: Sort) => history.push(setSortInUrl(location, sort));
    const handleFilter = (filter: Filter) => history.push(setFilterInUrl(location, filter));
    const handleNavigate = (labelID: string) => history.push(`/${labelID}`);

    // Move to the previous page if the current one becomes empty
    useEffect(() => {
        if (!loading && elements.length === 0 && page.page > 0) {
            handlePage(page.page - 1);
        }
    }, [loading]);

    /**
     * Put *IDs* to *checked* state
     * Uncheck others id *replace* is true
     */
    const handleCheck = (IDs: string[], checked: boolean, replace: boolean) =>
        setCheckedElements(
            elements.reduce((acc, { ID = '' }) => {
                const wasChecked = checkedElements[ID];
                const toCheck = IDs.includes(ID);
                acc[ID] = toCheck ? checked : replace ? !checked : wasChecked;
                return acc;
            }, {} as { [ID: string]: boolean })
        );

    const handleUncheckAll = () => handleCheck([], true, true);

    const showToolbar = !breakpoints.isNarrow || !elementID;
    const showList = columnMode || !elementID;
    const showContentPanel = (columnMode && !!expectedLength) || !!elementID;
    const showPlaceholder = !breakpoints.isNarrow && (!elementID || !!checkedIDs.length);
    const showContentView = showContentPanel && !!elementID;
    const elementIDForList = checkedIDs.length ? undefined : elementID;

    return (
        <>
            {showToolbar && (
                <Toolbar
                    location={location}
                    labelID={labelID}
                    elementID={elementID}
                    selectedIDs={selectedIDs}
                    elements={elements}
                    mailSettings={mailSettings}
                    columnMode={columnMode}
                    conversationMode={conversationMode}
                    breakpoints={breakpoints}
                    onCheck={handleCheck}
                    page={page}
                    onPage={handlePage}
                    sort={sort}
                    onSort={handleSort}
                    filter={filter}
                    onFilter={handleFilter}
                    onBack={handleBack}
                    onElement={handleElement}
                    onNavigate={handleNavigate}
                />
            )}
            <PrivateMainArea className="flex" hasToolbar={showToolbar} hasRowMode={!showContentPanel}>
                {showList && (
                    <List
                        location={location}
                        labelID={labelID}
                        loading={loading}
                        expectedLength={expectedLength}
                        columnLayout={columnLayout}
                        mailSettings={mailSettings}
                        elementID={elementIDForList}
                        elements={elements}
                        checkedIDs={checkedIDs}
                        onCheck={handleCheck}
                        onClick={handleElement}
                        userSettings={userSettings}
                        isSearch={isSearch}
                    />
                )}
                {showContentPanel && (
                    <section className="view-column-detail flex flex-column flex-item-fluid no-scroll">
                        {showPlaceholder && (
                            <PlaceholderView
                                welcomeFlag={welcomeFlag}
                                location={location}
                                labelID={labelID}
                                mailSettings={mailSettings}
                                elementID={elementID}
                                checkedIDs={checkedIDs}
                                onUncheckAll={handleUncheckAll}
                            />
                        )}
                        {showContentView &&
                            (conversationMode ? (
                                <ConversationView
                                    hidden={showPlaceholder}
                                    labelID={labelID}
                                    mailSettings={mailSettings}
                                    conversationID={elementID as string}
                                    onBack={handleBack}
                                    onCompose={onCompose}
                                />
                            ) : (
                                <MessageOnlyView
                                    hidden={showPlaceholder}
                                    labelID={labelID}
                                    mailSettings={mailSettings}
                                    messageID={elementID as string}
                                    onBack={handleBack}
                                    onCompose={onCompose}
                                />
                            ))}
                    </section>
                )}
            </PrivateMainArea>
        </>
    );
};

export default MailboxContainer;
