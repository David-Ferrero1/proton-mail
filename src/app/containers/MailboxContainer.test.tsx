import React from 'react';
import { act } from 'react-dom/test-utils';
import { render } from '@testing-library/react';
import { Location, History } from 'history';
import { noop } from 'proton-shared/lib/helpers/function';

import MailboxContainer from './MailboxContainer';
import { useElements as useElementsSource } from '../hooks/useElements';
import * as ToolbarSource from '../components/toolbar/Toolbar';

const useElements = useElementsSource as jest.Mock;
const Toolbar = (ToolbarSource as any) as {
    props: { selectedIDs: string[]; onCheck: (IDs: string[], checked: boolean, replace: boolean) => void };
};

jest.mock('../components/list/List', () => () => 'List');
jest.mock('../components/conversation/ConversationView', () => () => 'ConversationView');
jest.mock('../components/message/MessageOnlyView', () => () => 'MessageOnlyView');
jest.mock('../components/toolbar/Toolbar');
jest.mock('../components/view/PlaceholderView', () => () => 'PlaceholderView');

jest.mock('../hooks/useMailboxPageTitle', () => ({ useMailboxPageTitle: jest.fn() }));
jest.mock('../hooks/useElements', () => ({ useElements: jest.fn() }));

describe('MailboxContainer', () => {
    const labelID = 'labelID';
    const elementID = 'elementID';
    const emptyProps = {
        labelID,
        mailSettings: {},
        elementID,
        location: {} as Location,
        history: {} as History,
        onCompose: noop
    };
    const elements = [{ ID: '0' }, { ID: '1' }, { ID: '2' }, { ID: '3' }, { ID: '4' }];

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('Snapshots', () => {
        it('should show loader instead of list when elements loading', () => {
            useElements.mockReturnValue([labelID, [], true, jest.fn()]);
            const result = render(<MailboxContainer {...emptyProps} />);
            expect(result).toMatchSnapshot();
        });

        it('should show list when elements finish loading', () => {
            useElements.mockReturnValue([labelID, [], false, jest.fn()]);
            const result = render(<MailboxContainer {...emptyProps} />);
            expect(result).toMatchSnapshot();
        });
    });

    describe('Selections', () => {
        it('should select the right elements for replace false', () => {
            useElements.mockReturnValue([labelID, elements, false, jest.fn()]);
            render(<MailboxContainer {...emptyProps} />);
            act(() => Toolbar.props.onCheck(['1'], true, false));
            expect(Toolbar.props.selectedIDs).toEqual(['1']);
            act(() => Toolbar.props.onCheck(['2'], true, false));
            expect(Toolbar.props.selectedIDs).toEqual(['1', '2']);
        });

        it('should select the right elements for replace true', () => {
            useElements.mockReturnValue([labelID, elements, false, jest.fn()]);
            render(<MailboxContainer {...emptyProps} />);
            act(() => Toolbar.props.onCheck(['1'], true, true));
            expect(Toolbar.props.selectedIDs).toEqual(['1']);
            act(() => Toolbar.props.onCheck(['2'], true, true));
            expect(Toolbar.props.selectedIDs).toEqual(['2']);
        });
    });
});
