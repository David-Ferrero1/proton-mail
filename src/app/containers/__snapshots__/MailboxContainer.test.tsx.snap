// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailboxContainer Snapshots should show list when elements finish loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Toolbar
      <div
        class="main-area--withToolbar flex-item-fluid flex reset4print,false"
      >
        <div
          class="items-column-list scroll-if-needed scroll-smooth-touch"
        >
          List
        </div>
        <section
          class="view-column-detail p2 flex-item-fluid scroll-if-needed"
        >
          ConversationView
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    Toolbar
    <div
      class="main-area--withToolbar flex-item-fluid flex reset4print,false"
    >
      <div
        class="items-column-list scroll-if-needed scroll-smooth-touch"
      >
        List
      </div>
      <section
        class="view-column-detail p2 flex-item-fluid scroll-if-needed"
      >
        ConversationView
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MailboxContainer Snapshots should show loader instead of list when elements loading 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Toolbar
      <div
        class="main-area--withToolbar flex-item-fluid flex reset4print,false"
      >
        <div
          class="items-column-list scroll-if-needed scroll-smooth-touch"
        >
          <div
            class="flex flex-justify-center h100"
          >
            Loader
          </div>
        </div>
        <section
          class="view-column-detail p2 flex-item-fluid scroll-if-needed"
        >
          ConversationView
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    Toolbar
    <div
      class="main-area--withToolbar flex-item-fluid flex reset4print,false"
    >
      <div
        class="items-column-list scroll-if-needed scroll-smooth-touch"
      >
        <div
          class="flex flex-justify-center h100"
        >
          Loader
        </div>
      </div>
      <section
        class="view-column-detail p2 flex-item-fluid scroll-if-needed"
      >
        ConversationView
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
